cmd/outline-ss-server/main.go

```diff
@@ -129,50 +102,120 @@ func (s *SSServer) loadConfig(filename string) error {
-	for _, keyConfig := range config.Keys {
-		portChanges[keyConfig.Port] = 1
-		cipherList, ok := portCiphers[keyConfig.Port]
-		if !ok {
-			cipherList = list.New()
-			portCiphers[keyConfig.Port] = cipherList
-		}
-		cryptoKey, err := shadowsocks.NewEncryptionKey(keyConfig.Cipher, keyConfig.Secret)
+	uniqueCiphers := 0
+	addrs := make([]NetworkAddr, 0)
+	addrCiphers := make(map[NetworkAddr]*list.List) // Values are *List of *CipherEntry.
+
+	for _, legacyKeyServiceConfig := range config.Keys {
+		cryptoKey, err := shadowsocks.NewEncryptionKey(legacyKeyServiceConfig.Cipher, legacyKeyServiceConfig.Secret)
 		if err != nil {
-			return fmt.Errorf("failed to create encyption key for key %v: %w", keyConfig.ID, err)
+			return fmt.Errorf("failed to create encyption key for key %v: %w", legacyKeyServiceConfig.ID, err)
 		}
-		entry := service.MakeCipherEntry(keyConfig.ID, cryptoKey, keyConfig.Secret)
-		cipherList.PushBack(&entry)
-	}
-	for port := range s.ports {
-		portChanges[port] = portChanges[port] - 1
+		entry := service.MakeCipherEntry(legacyKeyServiceConfig.ID, cryptoKey, legacyKeyServiceConfig.Secret)
+		for _, network := range []string{"tcp", "udp"} {
+			addr := NetworkAddr{
+				network: network,
+				Host:    "::",
+				Port:    uint(legacyKeyServiceConfig.Port),
+			}
+			addrs = append(addrs, addr)
+			ciphers, ok := addrCiphers[addr]
+			if !ok {
+				ciphers = list.New()
+				addrCiphers[addr] = ciphers
+			}
+			ciphers.PushBack(&entry)
+		}
+		uniqueCiphers += 1
 	}
-	for portNum, count := range portChanges {
-		if count == -1 {
-			if err := s.removePort(portNum); err != nil {
-				return fmt.Errorf("failed to remove port %v: %w", portNum, err)
+
+	for _, serviceConfig := range config.Services {
+		ciphers := list.New()
+		type cipherKey struct {
+			cipher string
+			secret string
+		}
+		existingCiphers := make(map[cipherKey]bool)
+		for _, keyConfig := range serviceConfig.Keys {
+			key := cipherKey{keyConfig.Cipher, keyConfig.Secret}
+			if _, exists := existingCiphers[key]; exists {
+				logger.Debugf("encryption key already exists for ID=`%v`. Skipping.", keyConfig.ID)
+				continue
 			}
-		} else if count == +1 {
-			if err := s.startPort(portNum); err != nil {
-				return err
+			cryptoKey, err := shadowsocks.NewEncryptionKey(keyConfig.Cipher, keyConfig.Secret)
+			if err != nil {
+				return fmt.Errorf("failed to create encyption key for key %v: %w", keyConfig.ID, err)
 			}
+			entry := service.MakeCipherEntry(keyConfig.ID, cryptoKey, keyConfig.Secret)
+			ciphers.PushBack(&entry)
+			existingCiphers[key] = true
+		}
+		uniqueCiphers += ciphers.Len()
+
+		for _, listener := range serviceConfig.Listeners {
+			addr, err := ParseNetworkAddr(listener.Address)
+			if err != nil {
+				return fmt.Errorf("error parsing listener address `%s`: %v", listener.Address, err)
+			}
+			addrs = append(addrs, addr)
+			addrCiphers[addr] = ciphers
 		}
 	}
-	for portNum, cipherList := range portCiphers {
```

Comment:
Perhaps an easier way to solve the reuse of ports is to set service handlers on the ports, rather than set the listeners of the services.

I imagine code like this:
```go
  newListenerSet := oldListenerSet.Clone()
  for each new service {
    handler := NewHandler(serviceConfig)
    for each listenConfig {
      listener := newListenerSet.GetListener(listenConfig)
      listener.Sethandler(handler)
    }
  }

  oldListenerSet.Close() // old services shutdown automatically on listener close
```

Note the listener set clone. With clone, you can reuse listeners from the source. However, it should fail if you listen on the same address twice.